src/colmap/estimators/generalized_pose.cc

```diff
@@ -0,0 +1,306 @@
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Author: Johannes L. Schoenberger (jsch-at-demuc-dot-de)
+
+#include "colmap/estimators/generalized_pose.h"
+
+#include "colmap/estimators/bundle_adjustment.h"
+#include "colmap/estimators/cost_functions.h"
+#include "colmap/estimators/generalized_absolute_pose.h"
+#include "colmap/estimators/pose.h"
+#include "colmap/geometry/rigid3.h"
+#include "colmap/math/matrix.h"
+#include "colmap/optim/ransac.h"
+#include "colmap/optim/support_measurement.h"
+#include "colmap/scene/camera.h"
+#include "colmap/sensor/models.h"
+#include "colmap/util/misc.h"
+
+#include <Eigen/Core>
+#include <ceres/ceres.h>
+
+namespace colmap {
+namespace {
+
+bool LowerVector3d(const Eigen::Vector3d& v1, const Eigen::Vector3d& v2) {
+  if (v1.x() < v2.x()) {
+    return true;
+  } else if (v1.x() == v2.x()) {
+    if (v1.y() < v2.y()) {
+      return true;
+    } else if (v1.y() == v2.y()) {
+      return v1.z() < v2.z();
+    } else {
+      return false;
+    }
+  } else {
+    return false;
+  }
+}
+
+}  // namespace
+
+bool EstimateGeneralizedAbsolutePose(
+    const RANSACOptions& options,
+    const std::vector<Eigen::Vector2d>& points2D,
+    const std::vector<Eigen::Vector3d>& points3D,
+    const std::vector<size_t>& camera_idxs,
+    const std::vector<Rigid3d>& cams_from_rig,
+    const std::vector<Camera>& cameras,
+    Rigid3d* rig_from_world,
+    size_t* num_inliers,
+    std::vector<char>* inlier_mask) {
+  CHECK_EQ(points2D.size(), points3D.size());
+  CHECK_EQ(points2D.size(), camera_idxs.size());
+  CHECK_EQ(cams_from_rig.size(), cameras.size());
+  CHECK_GE(*std::min_element(camera_idxs.begin(), camera_idxs.end()), 0);
```

Comment:
This should probably also work:
```suggestion
  CHECK_GE(std::min(camera_idxs), 0);
```

Reply:
`std::min` only works for initializer lists. `std::ranges:min` works but requires C++ 20.